Plugins/ExaTrkX/include/Acts/Plugins/ExaTrkX/detail/ConnectedComponents.cuh

```diff
@@ -97,6 +119,42 @@ __global__ void labelConnectedComponents(std::size_t numEdges,
   } while (__syncthreads_or(changed));
 }
 
+/// Hooking-kernel for implementing the FastSV loop on the host
+template <typename TEdge, typename TLabel>
+__global__ void hookEdges(std::size_t numEdges, const TEdge *sourceEdges,
+                          const TEdge *targetEdges, const TLabel *labels,
+                          TLabel *labelsNext, int *globalChanged) {
+  int i = threadIdx.x + blockIdx.x * blockDim.x;
+  if (i >= numEdges) {
+    return;
+  }
+
+  bool changed = false;
+  hookEdgesImpl(i, sourceEdges, targetEdges, labels, labelsNext, changed);
+
+  if (__syncthreads_or(changed) && threadIdx.x == 0) {
```

Comment:
Be extremely careful using barriers in kernels which have some threads return early. A thread returning can cause a barrier to become a deadlock.

Reply:
Thanks, a very good point. I updated the code to not use early return anymore.